### PR TITLE
WebKit2GTK: Add wxWebViewConfiguration and child window support

### PR DESCRIPTION
**Pull request by TcT2k**
Add some features to the WebKit2GTK backend that where recently added to Edge and macOS

wxWebViewConfiguration
==

Add support for more in depth configuration before creating a web view. Allows to customize the data path and access to the underlying context.

Add wxWebViewWindowFeatures and event
==

This replaces the previously implemented `wxWebViewWindowInfo`.

It explicitly **breaks** the previous API to enable WebKitGTK
integration and to make the usage in application code less error prone.

A new event `wxEVT_WEBVIEW_NEWWINDOW_FEATURES` is added to allow
access to the window features and the child web view.

New sample handling code looks like this:
```C++
// Bind new window handler
m_webView->Bind(wxEVT_WEBVIEW_NEWWINDOW, [](wxWebViewEvent& evt) {
    if (evt.GetURL() == "http://badwebsite.com")
        evt.Veto(); // Disallow new window for badwebsite.com
    else
        evt.Skip(); // Allow new window for all other websites
});
 
// Bind new window features handler
m_webView->Bind(wxEVT_WEBVIEW_NEWWINDOW_FEATURES, [](wxWebViewEvent& evt) {
    // Get target window features
    wxWebViewWindowFeatures* features = evt.GetTargetWindowFeatures();
    // Create a top level window for the child web view
    wxWindow* win = new wxWindow(this, wxID_ANY, features->GetPosition(), features->GetSize());
    wxSizer* sizer = new wxBoxSizer(wxVERTICAL);
    win->SetSizer(sizer);
    // Get the child web view
    wxWebView* childWebView = features->GetChildWebView();
    // Place the child web view into the window
    childWebView->Create(win, wxID_ANY);
    sizer->Add(childWebView, 1, wxEXPAND);
}
```

### CODE REVIEW

File: include/wx/gtk/webview_webkit.h [hunk 1]
```diff
@@ -216,9 +212,18 @@ class WXDLLIMPEXP_WEBVIEW wxWebViewFactoryWebKit : public wxWebViewFactory
                               const wxSize& size = wxDefaultSize,
                               long style = 0,
                               const wxString& name = wxASCII_STR(wxWebViewNameStr)) override
-    { return new wxWebViewWebKit(parent, id, url, pos, size, style, name); }
+    { 
+        wxWebView* webView = new wxWebViewWebKit;
+        if (webView->Create(parent, id, url, pos, size, style, name))
+            return webView;
+        else
+            return nullptr;
```
Comment (vadz) [inline include/wx/gtk/webview_webkit.h:220]: Shouldn't we `delete webView` here (or, rather, use `std::unique_ptr<>` for it in the first place)?

File: src/common/webview.cpp [hunk 2]
```diff
@@ -86,6 +87,24 @@ wxString wxWebViewConfiguration::GetDataPath() const
...
+// wxWebViewWindowFeatures
+wxWebViewWindowFeatures::wxWebViewWindowFeatures(wxWebView * childWebView):
+    m_childWebViewWasUsed(false),
+    m_childWebView(childWebView)
+{ }
+
+wxWebViewWindowFeatures::~wxWebViewWindowFeatures()
+{
+    if (m_childWebView && !m_childWebViewWasUsed)
+        delete m_childWebView;
```
Comment (vadz) [inline src/common/webview.cpp:99]: I wonder if we could avoid this by making `GetChildWebView()` reset the pointer to `nullptr` after returning it? In fact, what about using `unique_ptr<>` here too and returning it by value from `GetChildWebView()` to make it absolutely clear that the caller takes its ownership?

File: src/gtk/webview_webkit2.cpp [hunk 3]
```diff
@@ -608,6 +682,80 @@ wxgtk_authorize_authenticated_peer_cb(GDBusAuthObserver *,
...
 
+//-----------------------------------------------------------------------------
+// wxWebViewConfigurationImplWebKit
+//-----------------------------------------------------------------------------
+class wxWebViewConfigurationImplWebKit : public wxWebViewConfigurationImpl
+{
+public:
+    wxWebViewConfigurationImplWebKit():
+        m_webContext(nullptr)
+#if WEBKIT_CHECK_VERSION(2, 10, 0)
```
Comment (vadz) [inline src/gtk/webview_webkit2.cpp:693]: Should we perhaps define some `HAVE_WEBKIT_WEBSITE_DATA_MANAGER` once instead of checking for the version multiple times? This would be a bit more readable, IMHO, and would make it easier to get rid of this check later when we drop support for the versions < 2.10.

File: src/gtk/webview_webkit2.cpp [hunk 4]
```diff
@@ -608,6 +682,80 @@ wxgtk_authorize_authenticated_peer_cb(GDBusAuthObserver *,
...
+        m_websiteDataManager(nullptr)
+#endif
+    { }
+
+#if WEBKIT_CHECK_VERSION(2, 10, 0)
+    wxString GetDataPath() const override
+    {
+        GetOrCreateContext();
+        if (m_websiteDataManager)
+            return webkit_website_data_manager_get_base_data_directory(m_websiteDataManager);
```
Comment (vadz) [inline src/gtk/webview_webkit2.cpp:704]: We probably need a run-time version check for 2.10 here (and when creating it below), otherwise an app compiled on a newer system will crash when running on an older one and I think we want to support this.

As I've just recently written in [this comment](https://github.com/wxWidgets/wxWidgets/issues/23367#issuecomment-1479377777), I think it would be useful to add `wx_check_webkit_version()` function for doing it, as WebKit itself doesn't seem to provide one.

**Review by vadz — COMMENTED** [review]
Thanks for your work on this!

The new API looks perfectly reasonable to me and it's great that it now works on all platforms.

I just have some minor comments about (not) deleting raw pointers, I think it would be nice to start following good C++ practices in wx code (better late than never, right) and use `unique_ptr` instead whenever possible.

Thanks again!

File: include/wx/gtk/webview_webkit.h (hunk 1 above)
Comment (TcT2k) [reply include/wx/gtk/webview_webkit.h:220]: Changed to use `std::unique_ptr`

File: src/gtk/webview_webkit2.cpp [hunk 5]
```diff
@@ -73,6 +74,79 @@ bool wxGetStringFromJSResult(WebKitJavascriptResult* js_result, wxString* output
...
+        webkit_window_properties_get_geometry(m_properties, &m_geometry);
+        // Treat 0 as -1 to indicate that the value is not set
+        if (m_geometry.width == 0)
+            m_geometry.width = -1;
+        if (m_geometry.height == 0)
+            m_geometry.height = -1;
+        if (m_geometry.x == 0)
+            m_geometry.x = -1;
+        if (m_geometry.y == 0)
+            m_geometry.y = -1;
```
[the remark this answers is not in the document]
Comment (TcT2k) [reply src/gtk/webview_webkit2.cpp:96]: While valid I think it's still the right thing to do. As in most cases it won't be set and it's more in line with the other backends (where we can actually check if a value is available).

File: src/common/webview.cpp (hunk 2 above)
Comment (TcT2k) [reply src/common/webview.cpp:99]: Using a `std::unique_ptr` would probably make sense. But I intentionally wanted `GetChildWebView()` to be callable multiple times by application code. E.g. it could make sense to do something with the webview before processing the rest of `wxWebViewWindowFeatures` in another method.

As this is a control the application code shouldn't have to worry to much about ownership as that is more or less automatic once it's added to a window.

File: include/wx/webview.h [hunk 6]
```diff
@@ -361,10 +365,11 @@ class WXDLLIMPEXP_WEBVIEW wxWebViewWindowInfo
 
     virtual bool ShouldDisplayScrollBars() const = 0;
 
-    virtual wxWebView* CreateChildWebView() = 0;
+protected:
+    mutable bool m_childWebViewWasUsed; // modified from const method
```
Comment (vadz) [inline include/wx/webview.h:369]: Sorry for continuing to nitpick, but `GetChildWebView()` is not actually const, which is why I was wondering why it was mutable in the first place. Did you intend to make it const, or is there some other method modifying this variable somewhere?

Comment (TcT2k) [reply include/wx/webview.h:369]: I see, I indeed intented it to be const. Will change that.

File: src/gtk/webview_webkit2.cpp (hunk 4 above)
Comment (vadz) [reply src/gtk/webview_webkit2.cpp:711]: Yes, it's called "lazy binding" and is the default under Linux (and the vast majority of the other Unix systems although historically there were some exceptions, I believe). This is also the default behaviour of `dlopen()` (`RTLD_LAZY` is the default, but `RTLD_NOW` can be used to force immediately resolving everything).

File: src/gtk/webview_webkit2.cpp [hunk 7]
```diff
@@ -38,6 +39,20 @@
...
+
+// Function to check webkit version at runtime
+bool wx_check_webkit_version(int major, int minor, int micro)
+{
+    const unsigned int version = webkit_get_major_version() * 10000 +
+                                 webkit_get_minor_version() * 100 +
+                                 webkit_get_micro_version();
+    const unsigned int required = major * 10000 + minor * 100 + micro;
+    return version >= required;
+}
```
Comment (swt2c) [inline src/gtk/webview_webkit2.cpp:54]: Can `wx_check_webkit_version` please be moved into a separate commit, and perhaps be merged separately ASAP?  I could use it for the other webkit2 changes.

Comment (TcT2k) [reply src/gtk/webview_webkit2.cpp:54]: I've added it as a separate commit. But this PR can probably be merged soon anyway and would include this change.

Comment (swt2c) [reply src/gtk/webview_webkit2.cpp:54]: Thanks - that will make it easier to backport - since all of the changes in this PR will probably not be backportable.

**Comment by vadz** [comment]
Thanks, merged now and, indeed, this won't be backported entirely but we can backport 42ca043.
